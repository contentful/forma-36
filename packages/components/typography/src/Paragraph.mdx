---
title: 'Paragraph'
type: 'component'
status: 'stable'
slug: /components/paragraph/
github: 'https://github.com/contentful/forma-36/tree/master/packages/components/typography/src'
storybook: 'https://v4-f36-storybook.netlify.app/?path=/story/typography-paragraph--basic'
typescript: ./Paragraph.tsx
---

`Paragraph` component is rendered as `<p>` and with default bottom margin of `16px` (`spacingM`) applied. It is one of the most commonly used typography components.

### Import

```js static=true
import { Paragraph } from '@contentful/f36-components';
// or
import { Paragraph } from '@contentful/f36-typography';
```

## Code examples

```jsx
<Paragraph>I love Forma 36 design system.</Paragraph>
```

You can override the default margin bottom by specifying it explicitly.

```jsx
<Flex flexDirection="column">
  <Paragraph marginBottom="none">I love Forma 36 design system.</Paragraph>
  <Paragraph marginBottom="spacingXl">I love Forma 36 design system.</Paragraph>
  <Paragraph marginTop="spacingS">I love Forma 36 design system.</Paragraph>
</Flex>
```

## Props

import { Props } from '@contentful/f36-docs-utils';

<Props of="Paragraph" />
