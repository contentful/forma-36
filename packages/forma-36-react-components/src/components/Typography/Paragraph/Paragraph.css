@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.Paragraph {
  display: block;
  font-family: var(--font-stack-primary);
  font-weight: var(--font-weight-normal);
  color: var(--color-text-mid);
  font-size: var(--font-size-m);
  line-height: var(--line-height-m);
  margin: 0;
}
