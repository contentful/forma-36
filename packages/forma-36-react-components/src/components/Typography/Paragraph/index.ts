export * from './Paragraph';
export { default } from './Paragraph';
