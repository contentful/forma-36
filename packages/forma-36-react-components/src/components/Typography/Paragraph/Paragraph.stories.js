import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, selectV2 } from '@storybook/addon-knobs/react';
import { host } from 'storybook-host';
import { withInfo } from '@storybook/addon-info';

import Paragraph from './Paragraph';

storiesOf('Components|Typography/Paragraph', module)
  .addDecorator(
    host({
      align: 'center middle',
      cropMarks: false,
    }),
  )
  .add(
    'default',
    withInfo()(() => (
      <Paragraph
        extraClassNames={text('Extra Class Names', '')}
        element={selectV2(
          'Element',
          ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p'],
          'h1',
        )}
      >
        {text('Content', 'Paragraph')}
      </Paragraph>
    )),
  );
