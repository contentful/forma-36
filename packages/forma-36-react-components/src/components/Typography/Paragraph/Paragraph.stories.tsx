import React from 'react';
import Paragraph, { ParagraphProps } from './Paragraph';

export default {
  title: 'Typography/Paragraph',
  component: Paragraph,
  parameters: {
    propTypes: [Paragraph['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
  },
};

interface Args extends ParagraphProps {
  paragraphText: string;
}

export const Basic = ({ paragraphText, ...args }: Args) => (
  <Paragraph {...args}>{paragraphText}</Paragraph>
);

Basic.args = {
  paragraphText: 'Paragraph',
};
