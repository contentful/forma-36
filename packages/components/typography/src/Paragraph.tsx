import React, { useContext } from 'react';
import tokens from '@contentful/f36-tokens';
import { css, cx } from 'emotion';
import {
  PolymorphicComponent,
  PolymorphicComponentWithRef,
} from '@contentful/f36-core';
import type { HeadingInternalProps, HeadingProps } from './Heading';
import { Heading } from './Heading';
import { TypographyContext } from './Typography';

const DEFAULT_TAG = 'p';

export type ParagraphProps<E extends React.ElementType> = HeadingProps<E>;

const _Paragraph: PolymorphicComponentWithRef<
  HeadingInternalProps,
  typeof DEFAULT_TAG
> = ({ children, className, ...otherProps }, ref) => {
  const configuration = useContext(TypographyContext);
  return (
    <Heading
      as={DEFAULT_TAG}
      testId="cf-ui-paragraph"
      marginBottom={configuration.paragraph}
      className={cx(
        css({
          fontWeight: tokens.fontWeightNormal,
          color: tokens.colorTextMid,
          fontSize: tokens.fontSizeM,
          lineHeight: tokens.lineHeightM,
        }),
        className,
      )}
      {...otherProps}
      ref={ref}
    >
      {children}
    </Heading>
  );
};

export const Paragraph: PolymorphicComponent<
  HeadingInternalProps,
  typeof DEFAULT_TAG
> = React.forwardRef(_Paragraph);
