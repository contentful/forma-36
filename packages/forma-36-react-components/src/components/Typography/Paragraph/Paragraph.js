import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './Paragraph.css';

import { TypographyContext } from '../Typography/Typography';

export class Paragraph extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    element: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p']),
  };

  static defaultProps = {
    element: 'p',
    extraClassNames: undefined,
    testId: 'cf-ui-paragraph',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      element,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.Paragraph, extraClassNames);

    const Element = element;

    return (
      <TypographyContext.Consumer>
        {({ paragraph }) => {
          return (
            <Element
              className={cn(classNames, [
                paragraph && `f36-margin-bottom--${paragraph.spacing}`,
              ])}
              data-test-id={testId}
              {...otherProps}
            >
              {children}
            </Element>
          );
        }}
      </TypographyContext.Consumer>
    );
  }
}

export default Paragraph;
