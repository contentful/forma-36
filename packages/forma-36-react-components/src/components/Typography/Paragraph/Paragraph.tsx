import React from 'react';
import cn from 'classnames';
import styles from './Paragraph.css';

import { TypographyContext } from '../Typography';

export interface ParagraphProps {
  element?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';
  className?: string;
  children?: React.ReactNode;
  testId?: string;
  style?: React.CSSProperties;
}

export function Paragraph({
  className,
  children,
  testId = 'cf-ui-paragraph',
  element = 'p',
  ...otherProps
}: ParagraphProps): React.ReactElement {
  const classNames = cn(styles['Paragraph'], className);

  const Element = element!; // eslint-disable-line @typescript-eslint/no-non-null-assertion

  return (
    <TypographyContext.Consumer>
      {(value) => {
        return (
          <Element
            className={cn(classNames, [
              value['paragraph'] &&
                `f36-margin-bottom--${value['paragraph']['spacing']}`,
            ])}
            data-test-id={testId}
            {...otherProps}
          >
            {children}
          </Element>
        );
      }}
    </TypographyContext.Consumer>
  );
}

export default Paragraph;
