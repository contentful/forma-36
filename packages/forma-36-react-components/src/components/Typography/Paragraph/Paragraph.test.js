import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import Paragraph from './Paragraph';

it('renders the component', () => {
  const output = shallow(<Paragraph>Paragraph</Paragraph>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <Paragraph extraClassNames="my-extra-class">Paragraph</Paragraph>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component h3', () => {
  const output = shallow(<Paragraph element="h3">Paragraph</Paragraph>);

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(<Paragraph>Paragraph</Paragraph>).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
