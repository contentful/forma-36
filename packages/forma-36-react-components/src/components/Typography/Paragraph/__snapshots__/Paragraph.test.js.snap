// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<p
  className="Paragraph"
  data-test-id="cf-ui-paragraph"
>
  Paragraph
</p>
`;

exports[`renders the component h3 1`] = `
<h3
  className="Paragraph"
  data-test-id="cf-ui-paragraph"
>
  Paragraph
</h3>
`;

exports[`renders the component with an additional class name 1`] = `
<p
  className="Paragraph my-extra-class"
  data-test-id="cf-ui-paragraph"
>
  Paragraph
</p>
`;
